kotsadm/pkg/dex/kotsadm.go

```diff
@@ -0,0 +1,224 @@
+	}
+
+	return &config, nil
+}
+
+func GetKotsadmOIDCProvider() (*oidc.Provider, error) {
+	dexConfig, err := GetKotsadmDexConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get kotsadm dex config")
+	}
+
+	ctx := oidc.ClientContext(context.Background(), http.DefaultClient)
+	provider, err := oidc.NewProvider(ctx, dexConfig.Issuer)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to query provider %q", dexConfig.Issuer)
+	}
+
+	return provider, nil
+}
+
+func GetKotsadmOAuth2Config() (*oauth2.Config, error) {
+	dexConfig, err := GetKotsadmDexConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get kotsadm dex config")
+	}
+
+	provider, err := GetKotsadmOIDCProvider()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get kotsadm oidc provider")
+	}
+
+	var kotsadmClient *dexstorage.Client
+	for _, client := range dexConfig.StaticClients {
+		if client.ID == "kotsadm" {
+			kotsadmClient = &client
+			break
+		}
+	}
+	if kotsadmClient == nil {
+		return nil, errors.New("kotsadm dex client not found")
+	}
+
+	oauth2Config := oauth2.Config{
+		ClientID:     kotsadmClient.ID,
+		ClientSecret: kotsadmClient.Secret,
+		Endpoint:     provider.Endpoint(),
+		Scopes:       getScopes(),
+		RedirectURL:  kotsadmClient.RedirectURIs[0],
+	}
+
+	return &oauth2Config, nil
+}
+
+func getScopes() []string {
+	// TODO "offline_access"
+	return []string{"openid", "profile", "email", "groups"}
+}
+
+func SetDexState(state string) error {
+	stateMtx.Lock()
+	defer stateMtx.Unlock()
+
+	secret := stateSecretResource(DexStateSecretName, state)
+
+	clientset, err := k8s.Clientset()
+	if err != nil {
+		return errors.Wrap(err, "failed to get k8s client set")
+	}
+
+	existingSecret, err := clientset.CoreV1().Secrets(os.Getenv("POD_NAMESPACE")).Get(context.TODO(), DexStateSecretName, metav1.GetOptions{})
+	if err != nil {
+		if !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing dex state secret")
+		}
+
+		_, err = clientset.CoreV1().Secrets(os.Getenv("POD_NAMESPACE")).Create(context.TODO(), secret, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrap(err, "failed to create dex state secret")
+		}
+
+		return nil
+	}
+
+	existingSecret = updateStateSecret(existingSecret, secret)
+
+	_, err = clientset.CoreV1().Secrets(os.Getenv("POD_NAMESPACE")).Update(context.TODO(), existingSecret, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to update dex state secret")
+	}
+
+	return nil
+}
+
+func GetDexState(state string) (string, error) {
+	clientset, err := k8s.Clientset()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get k8s client set")
+	}
+
+	secret, err := clientset.CoreV1().Secrets(os.Getenv("POD_NAMESPACE")).Get(context.TODO(), DexStateSecretName, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get dex state secret")
+	}
+
+	return string(secret.Data[state]), nil
```

Comment:
should this check expiration before returning?

Reply:
we've added expiration just to cleanup the secret, not to prevent logging in. that might be something that is built into dex/oidc.